fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -1155,6 +1155,67 @@ void determineCommittedTransactions(CommitBatchContext* self) {
 	}
 }
 
+inline int64_t extractTenantIdFromStringRef(StringRef s) {
+	if (s.size() < TenantAPI::PREFIX_SIZE) {
+		return TenantInfo::INVALID_TENANT;
+	}
+	// Parse mutation key to determine mutation encryption domain
+	StringRef prefix = s.substr(0, TenantAPI::PREFIX_SIZE);
+	return TenantAPI::prefixToId(prefix, EnforceValidTenantId::False);
+}
+
+int64_t extractTenantIdFromSingleKeyMutation(MutationRef m) {
+	ASSERT(!isSystemKey(m.param1));
+	ASSERT(isSingleKeyMutation((MutationRef::Type)m.type));
+
+	return extractTenantIdFromStringRef(m.param1);
+}
+
+// Return true if a single-key mutation is associated with a valid tenant id or a system key
+// FIXME(xwang): handle clear range properly
+bool validTenantAccess(MutationRef m) {
+	if (isSystemKey(m.param1))
+		return true;
+
+	if (isSingleKeyMutation((MutationRef::Type)m.type)) {
+		auto tenantId = extractTenantIdFromSingleKeyMutation(m);
+		// throw exception for invalid raw access
+		if (tenantId == TenantInfo::INVALID_TENANT) {
+			return false;
+		}
+	} else {
+		// For clear range, we allow raw access
+		ASSERT_EQ(m.type, MutationRef::Type::ClearRange);
+		auto beginTenantId = extractTenantIdFromStringRef(m.param1);
+		auto endTenantId = extractTenantIdFromStringRef(m.param2);
+		CODE_PROBE(beginTenantId != endTenantId, "Clear Range raw access or cross multiple tenants");
+	}
+	return true;
+}
+
+// Return true if all tenant check pass. Otherwise, return false and send error back
+bool validTenantAccess(const CommitTransactionRequest& tr, ProxyCommitData* const pProxyCommitData) {
+	bool isValid = checkTenant(pProxyCommitData, tr.tenantInfo.tenantId, tr.tenantInfo.name);
+
+	if (!isValid) {
+		tr.reply.sendError(unknown_tenant());
+		return false;
+	}
+
+	// only do the mutation check when the transaction use raw_access option and the tenant mode is required
+	if (pProxyCommitData->getTenantMode() != TenantMode::REQUIRED) {
```

Comment:
Should this also include a condition for if the transaction specified a tenant? That way we avoid looping over the mutations for normal tenant transactions.